### PR TITLE
Specify most of `Option`'s methods

### PR DESCRIPTION
**Pull request by arnaudgolfouse**
A lot of methods involving closures are a bit awkward to use right now, this should get better with closure automatic spec inference.

### CODE REVIEW

File: creusot-contracts/src/std/option.rs [hunk 1]
```diff
@@ -35,63 +35,276 @@ extern_spec! {
...
+                    F: FnOnce(&T);
+
+                #[requires(match self {
+                    None => true,
+                    Some(t) => f.precondition((t,)),
+                })]
+                #[ensures(match self {
+                    None => result == default,
+                    Some(t) => f.postcondition_once((t,), result)
+                })]
```
Comment (xldenis) [inline creusot-contracts/src/std/option.rs:143]: These kinds of specs really make me wish for a predicate-transformer style specification, or some alternate mechanism to reduce this duplication.

**Review by xldenis — APPROVED** [review]
The specifications _look_ good when reading them but they are quite repetitive and its always hard to be 100% certain when adding axioms like this. 

Two thoughts occur to me: 
1. It would be nice to have better syntax than `postcondition_mut` & `precondition` to describe closure calls (even if it desugars into that). 
2. To alleviate the concerns of these axioms, and potentially allowing the "inference" of specifications for these simple functions, we could have an "external bodies" mechanism in `extern_specs!` where we actually provide the body and prove the spec (if provided) holds. Creusot would then treat this as the actual body for the relevant function.

**Comment by arnaudgolfouse** [comment]
Hard agree on point 2: I actually did this by hand for some of those function where the spec wasn't entirely clear (e.g. `filter` and `take_if`)

**Comment by arnaudgolfouse** [comment]
Also, I could not write a spec for `unwrap_unchecked` because this is an `unsafe` function.

**Comment by xldenis** [comment]
> Also, I could not write a spec for unwrap_unchecked because this is an unsafe function.

Wait why is this relevant? We can definitely write the spec, its the same as `unwrap`. Or is it a limitation in the `extern_spec!` macro?

File: creusot-contracts/src/std/option.rs [hunk 2]
```diff
@@ -35,63 +35,276 @@ extern_spec! {
...
+                        (predicate.postcondition_once((&t,), true) != predicate.postcondition_once((&t,), false)),
+                })]
+                #[ensures(match self {
+                    None => result == None,
+                    Some(t) => if predicate.postcondition_once((&t,), true) {
+                        result == Some(t)
+                    } else {
+                        result == None && resolve(&t)
+                    },
+                })]
```
Comment (jhjourdan) [inline creusot-contracts/src/std/option.rs:214]: Shorter spec, which does not requires determinism of the predicate:


```suggestion
                #[ensures(match self {
                    None => result == None,
                    Some(t) =>
                           predicate.postcondition_once((&t,), true) && result == Some(t)
                        || predicate.postcondition_once((&t,), false) && result == None && resolve(&t)
                    },
                })]
```

Comment (jhjourdan) [reply creusot-contracts/src/std/option.rs:214]: Or even:

```suggestion
                #[ensures(match self {
                    None => result == None,
                    Some(t) => match result {
                        None => predicate.postcondition_once((&t,), false) && resolve(&t),
                        Some(r) => predicate.postcondition_once((&t,), true) && r == t
                    },
                })]
```

File: creusot-contracts/src/std/option.rs [hunk 3]
```diff
@@ -35,63 +35,276 @@ extern_spec! {
...
+                #[requires(match *self {
+                    None => true,
+                    Some(t) => forall<b:&mut T> *b == t ==> predicate.precondition((b,)),
+                })]
+                #[ensures(match *self {
+                    None => result == None && ^self == None,
+                    Some(cur) =>
+                        exists<b: &mut T, res: bool> cur == *b && predicate.postcondition_once((b,), res) &&
+                            if res {
+                                ^self == None && result == Some(^b)
```
Comment (jhjourdan) [inline creusot-contracts/src/std/option.rs:279]: BUG:

```suggestion
                                ^self == None && result == Some(*b)
```

Comment (jhjourdan) [reply creusot-contracts/src/std/option.rs:279]: Well, no, that's good, sorry!

File: creusot-contracts/src/std/option.rs [hunk 4]
```diff
@@ -35,63 +35,276 @@ extern_spec! {
...
+                where
+                    F: FnOnce() -> T;
+
                 #[pure]
                 #[ensures(result == *self && ^self == None)]
                 fn take(&mut self) -> Option<T>;
 
+                #[requires(match *self {
+                    None => true,
+                    Some(t) => forall<b:&mut T> *b == t ==> predicate.precondition((b,)),
```
Comment (jhjourdan) [inline creusot-contracts/src/std/option.rs:272]:
```suggestion
                    Some(t) => forall<b:&mut T> inv(b) && *b == t ==> predicate.precondition((b,)),
```

**Comment by arnaudgolfouse** [comment]
The `unwrap_unchecked` method is now specified, and I added an `extern_spec` body to the "difficult" functions (the ones with closures)

**Pull request merged**